BUG: Fix issues caused by invalid Segmentation binary labelmap scalar types

Sunderlandkyl opened the pull request:
Adds the functions `ValidateSegmentationImageType` and `CastToSmallestUnsignedIntegerType` to vtkOrientedImageDataResample.
- `IsImageScalarTypeValid`: Returns True if the scalar type of the image is valid for representing segmentations. False otherwise.
- `CastToSmallestUnsignedIntegerType`: Casts the contents of the image to the smallest unsigned integer type that can contain all values in the image.
- `ValidateSegmentationImageType`: Checks to see if the scalar type of the image is a valid type, and casts the image to the smallest valid type if it is not.

The function `ValidateSegmentationImageType` is used during binary labelmap read/import to automatically convert the labelmap.
If a labelmap representation with an invalid scalar type is added to a segment, we only log a warning and don't automatically convert the labelmap.
Before performing binary labelmap to closed surface conversion, we check that the scalar type is valid, and log/return an error if it is not.

This should prevent errors from arising due to the use of floating point images for segmentations.

Re #6941

pieper commented on Libs/vtkSegmentationCore/vtkOrientedImageDataResample.cxx:
```diff
@@ -1835,9 +1835,100 @@ bool vtkOrientedImageDataResample::IsLabelInMask(
+    return false;
+  }
+
+  vtkIdType validScalarTypes[3] =
+  {
+    VTK_UNSIGNED_CHAR,
+    VTK_UNSIGNED_SHORT,
+    VTK_UNSIGNED_INT
+    //VTK_UNSIGNED_LONG // See issue #5427.
+                        // vtkImageThreshold is not able to fill thresholded images with VTK_LONG_MAX, preventing segment erase.
+  };
+
+  int scalarType = image->GetScalarType();
+  for (int i = 0; i < 3; i++)
```
It would be better not to have the hard-coded number 3 here.  Probably an `#define` would be better.

Sunderlandkyl replied:
Actually, since validScalarTypes is defined locally it probably make sense to use:
```
for (vtkIdType validScalarType : validScalarTypes)
```

Unless we should define the list of valid scalar types elsewhere.

pieper commented on Libs/vtkSegmentationCore/vtkOrientedImageDataResample.cxx:
```diff
@@ -1835,9 +1835,100 @@ bool vtkOrientedImageDataResample::IsLabelInMask(
+  }
+  else
+  {
+    vtkErrorWithObjectMacro(image, "CastImageToSmallestIntegerType: Scalar range is too large to fit into an integer type");
```
We could enable `ClampOverflow` in `vtkImageCast` and make this a warning rather than an error.  It might be more useful to the user to have something load rather than fail.  Or make it an option if we don't want to lose data with only a warning.

Sunderlandkyl replied:
Ok, we now clamp overflow and log a warning.

pieper commented on Libs/vtkSegmentationCore/vtkOrientedImageDataResample.cxx:
```diff
@@ -1835,9 +1835,100 @@ bool vtkOrientedImageDataResample::IsLabelInMask(
+    return false;
+  }
+
+  if (scalarType == image->GetScalarType())
```
Shouldn't we also check that any floating point data is losslessly castable to integer and warn if not?  Otherwise we would be silently losing data.  I know at least some of the float data that a user tried to load as a segmentation recently had non-integral values.

Sunderlandkyl replied:
We don't check specifically for lossless castability, however we always log a warning when casting from a floating point type to an integer type so that the user knows that it is a possibility.

The non-integral values are truncated during cast, which is consistent with how they are displayed in labelmap nodes.

pieper reviewed (commented):
Thanks for working on this 👍 Just a few comments/questions.

lassoan commented on Libs/vtkSegmentationCore/vtkOrientedImageDataResample.h:
```diff
@@ -183,10 +183,25 @@ class vtkSegmentationCore_EXPORT vtkOrientedImageDataResample : public vtkObject
+  /// Validate the segmentation image type, casting to another type if necessary.
+  /// \param image Image to validate
+  /// \return True if the image scalar type is valid, or is cast to a valid type, false otherwise
+  static bool ValidateSegmentationImageType(vtkImageData* image);
```
This method performs two quite simple operations, which is probably not used that often (and it makes error reporting a bit more complicated). So, I think we probably don't need this.

lassoan commented on Libs/vtkSegmentationCore/vtkOrientedImageDataResample.cxx:
```diff
@@ -1835,9 +1835,125 @@ bool vtkOrientedImageDataResample::IsLabelInMask(
 }
 
+//----------------------------------------------------------------------------
+bool vtkOrientedImageDataResample::IsSegmentationScalarTypeValid(vtkImageData* image)
```
It is a bit complex that the scalar types are checked in two slightly different ways (one is segmentation-specific as it excludes long int and float types, the other is somewhat-segmentation-specific as it excludes long int types).

It would be better to have just one segmentation-specific check, which based on the scalar typa and range tells if the the image type is TYPE_OK, TYPE_CONVERSION_NEEDED, TYPE_CONVERSION_TRUNCATION_NEEDED, TYPE_ERROR. Depending on this, warning or error is reported and/or logged and casting is called. Casting does not need to do any warnings (it is a casting function, so the developer must know that casting may cause data loss; but maybe in the future if we need it then we could add an enum return value to distinguish between lossless and potentially lossy conversion).

lassoan reviewed (commented):
Thank you, it looks very nice and it is great that you added a specific test for it. We could simplify the checking and error/warning reporting a bit - see my inline comments.

lassoan commented on Modules/Loadable/Segmentations/Logic/vtkSlicerSegmentationsModuleLogic.cxx:
```diff
@@ -1441,6 +1441,34 @@ bool vtkSlicerSegmentationsModuleLogic::ImportLabelmapToSegmentationNode(vtkMRML
+      case vtkOrientedImageDataResample::TYPE_ERROR:
+        vtkGenericWarningMacro("vtkMRMLSegmentationStorageNode::ImportLabelmapToSegmentationNode: "
+          "Failed to cast image to integer type.");
+        return 0;
```
```suggestion
        return false;
```

lassoan commented on Libs/MRML/Core/vtkMRMLSegmentationStorageNode.cxx:
```diff
@@ -694,6 +694,25 @@ int vtkMRMLSegmentationStorageNode::ReadBinaryLabelmapRepresentation(vtkMRMLSegm
+  {
+    case vtkOrientedImageDataResample::TYPE_CONVERSION_TRUNCATION_NEEDED:
+      vtkWarningToMessageCollectionMacro(this->GetUserMessages(), "vtkMRMLSegmentationStorageNode::ReadBinaryLabelmapRepresentation",
+        "Segmentation is a floating point scalar type and will be cast to an integer type. Voxel values may be truncated");
```
I don't see where the values are actually truncated or clamped.

lassoan commented on Modules/Loadable/Segmentations/Logic/vtkSlicerSegmentationsModuleLogic.cxx:
```diff
@@ -1441,6 +1441,34 @@ bool vtkSlicerSegmentationsModuleLogic::ImportLabelmapToSegmentationNode(vtkMRML
+    switch (ret)
+    {
+      case vtkOrientedImageDataResample::TYPE_CONVERSION_TRUNCATION_NEEDED:
+        vtkGenericWarningMacro(
```
It would be nice to able to return this as userMessages to allow the error message to be displayed in the GUI.

lassoan commented on Libs/MRML/Core/vtkMRMLSegmentationStorageNode.cxx:
```diff
@@ -694,6 +694,25 @@ int vtkMRMLSegmentationStorageNode::ReadBinaryLabelmapRepresentation(vtkMRMLSegm
+        "Segmentation is a floating point scalar type and will be cast to an integer type. Voxel values may be truncated");
+      break;
+    case vtkOrientedImageDataResample::TYPE_CONVERSION_CLAMPING_NEEDED:
+       vtkWarningToMessageCollectionMacro(this->GetUserMessages(), "vtkMRMLSegmentationStorageNode::ReadBinaryLabelmapRepresentation",
```
It would be nice to able to return this as userMessages to allow the error message to be displayed in the GUI.

Sunderlandkyl replied:
Isn't this one returned already?

lassoan reviewed (commented):
I don't see where the casting of float segmentation file to int happens and based on my testing it seems that it does not get converted. Please check!

I also added a commit that returns warning messages from the logic, which allows the warning to be displayed to the user. Maybe we will decide later to not display this warning to the user, to avoid unnecessary confusion, but it may worth giving it a try to educate users that float segmentations are not good.

jcfr commented on Libs/vtkSegmentationCore/vtkOrientedImageDataResample.h:
```diff
@@ -183,10 +183,39 @@ class vtkSegmentationCore_EXPORT vtkOrientedImageDataResample : public vtkObject
+    TYPE_CONVERSION_TRUNCATION_NEEDED,
+    TYPE_CONVERSION_CLAMPING_NEEDED,
+    TYPE_ERROR
+  };
```
```suggestion
  };

```

jcfr commented on Libs/vtkSegmentationCore/vtkOrientedImageDataResample.h:
```diff
@@ -183,10 +183,39 @@ class vtkSegmentationCore_EXPORT vtkOrientedImageDataResample : public vtkObject
   static bool IsLabelInMask(vtkOrientedImageData* binaryLabelmap, vtkOrientedImageData* mask,
     int extent[6]=nullptr, int maskThreshold=0);
 
+  enum
```
```suggestion
  enum ImageTypeCheckResult
```

jcfr commented on Libs/vtkSegmentationCore/vtkOrientedImageDataResample.h:
```diff
@@ -183,10 +183,39 @@ class vtkSegmentationCore_EXPORT vtkOrientedImageDataResample : public vtkObject
+    TYPE_CONVERSION_CLAMPING_NEEDED,
+    TYPE_ERROR
+  };
+  /// Determine if the segmentation image scalar type is valid for representing segmentations
```
```suggestion
  /// Determine if the scalar type of the image is valid for representing segmentations.
```

jcfr commented on Libs/vtkSegmentationCore/vtkOrientedImageDataResample.h:
```diff
@@ -183,10 +183,39 @@ class vtkSegmentationCore_EXPORT vtkOrientedImageDataResample : public vtkObject
+  };
+  /// Determine if the segmentation image scalar type is valid for representing segmentations
+  /// \param image Image to validate
+  /// \return Value representing if the image scalar type is valid, and if not, what kind of conversion is needed
```
```
  /// \return ImageTypeCheckResult value representing if the image scalar type is valid, and if not, what kind of conversion is needed.
  /// \sa vtkImageData::GetScalarType()
```

jcfr commented on Libs/vtkSegmentationCore/vtkOrientedImageDataResample.h:
```diff
@@ -183,10 +183,39 @@ class vtkSegmentationCore_EXPORT vtkOrientedImageDataResample : public vtkObject
+  /// Determine if the segmentation image scalar type is valid for representing segmentations
+  /// \param image Image to validate
+  /// \return Value representing if the image scalar type is valid, and if not, what kind of conversion is needed
+  static int IsSegmentationScalarTypeValid(vtkImageData* image);
```
```suggestion
  static int IsImageScalarTypeValid(vtkImageData* image);
``

jcfr commented on Libs/vtkSegmentationCore/vtkOrientedImageDataResample.h:
```diff
@@ -183,10 +183,39 @@ class vtkSegmentationCore_EXPORT vtkOrientedImageDataResample : public vtkObject
+  /// \return Value representing if the image scalar type is valid, and if not, what kind of conversion is needed
+  static int IsSegmentationScalarTypeValid(vtkImageData* image);
+
+  /// Determine the smallest integer type that can contain the specified scalar range
+  /// \param scalarRange Range of the scalar values that should be representable by the image data type
+  /// \return Smallest integer type that can contain the specified scalar range
```
```suggestion
  /// Determine the smallest integer type that can contain the specified scalar range.
  /// \param scalarRange Range of the scalar values that should be representable by the image data type
  /// \return Smallest integer type that can contain the specified scalar range. Possible values for are VTK_UNSIGNED_CHAR, VTK_CHAR, VTK_UNSIGNED_SHORT, VTK_SHORT, VTK_UNSIGNED_INT, VTK_INT and -1 if no valid integer type can contain the scalar range.
```

Sunderlandkyl committed:
BUG: Fix issues caused by invalid Segmentation binary labelmap scalar types

Adds the functions ValidateSegmentationImageType and CastToSmallestUnsignedIntegerType to vtkOrientedImageDataResample.
- IsImageScalarTypeValid: Returns True if the scalar type of the image is valid for representing segmentations. False otherwise.
- CastToSmallestUnsignedIntegerType: Casts the contents of the image to the smallest unsigned integer type that can contain all values in the image.
- ValidateSegmentationImageType: Checks to see if the scalar type of the image is a valid type, and casts the image to the smallest valid type if it is not.

The function ValidateSegmentationImageType is used during binary labelmap read/import to automatically convert the labelmap.
If a labelmap representation with an invalid scalar type is added to a segment, we only log a warning and don't automatically convert the labelmap.
Before performing binary labelmap to closed surface conversion, we check that the scalar type is valid, and log/return an error if it is not.

This should prevent errors from arising due to the use of floating point images for segmentations.

Re #6941

Co-authored-by: Jean-Christophe Fillion-Robin <jchris.fillionr@kitware.com>

jcfr commented on Libs/vtkSegmentationCore/vtkOrientedImageDataResample.h:
```diff
@@ -183,10 +183,41 @@ class vtkSegmentationCore_EXPORT vtkOrientedImageDataResample : public vtkObject
+
+  /// Determine the smallest integer type that can contain the specified scalar range.
+  /// \param scalarRange Range of the scalar values that should be representable by the image data type
+  /// \return Smallest integer type that can contain the specified scalar range. Possible values for are VTK_UNSIGNED_CHAR, VTK_CHAR, VTK_UNSIGNED_SHORT, VTK_SHORT, VTK_UNSIGNED_INT, VTK_INT and -1 if no valid integer type can contain the scalar range.
```
```suggestion
  /// \return Smallest integer type that can contain the specified scalar range. Possible values are VTK_UNSIGNED_CHAR, VTK_CHAR, VTK_UNSIGNED_SHORT, VTK_SHORT, VTK_UNSIGNED_INT, VTK_INT and -1 if no valid integer type can contain the scalar range.
```

jcfr reviewed (commented):
Thanks @Sunderlandkyl  , after addressing that last nitpick :+1:

lassoan commented:
Nice work, thank you @Sunderlandkyl
